### PR TITLE
Laboratory work #3, Margarita Velmisova - 22FPL1

### PR DESCRIPTION
**Pull request by velgarita**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +44,28 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        tokens = []
+        for word in text_words:
+            len_prev = len(tokens)
+            for alpha in word:
+                if alpha.isalpha():
+                    tokens.append(alpha)
+            if len(tokens) > len_prev:
+                tokens.append(self._end_of_word_token)
+
+        tokens = tokens[:-1]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:60]: Why are you doing this?

Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:60]: changed it

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -76,6 +109,14 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int):
+            return None
+
+        storage_invert = {ident: token for token, ident in self._storage.items()}
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:115]: changed it

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -104,6 +161,14 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+
+        if element not in self._storage:
+            ident = len(self._storage)
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:168]: I really don't need it, fixed this

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -129,6 +206,15 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if not isinstance(content, dict) or len(content) == 0:
+            return None
+
+        for n_gram in content['freq']:
+            for token in n_gram:
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:213]: ok

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -159,6 +256,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+
+        decoded_text = ''.join(decoded_corpus)
+        decoded_text = decoded_text.replace('_', ' ')
+        decoded_text = decoded_text.capitalize()
+        if decoded_text[-1] == ' ':
+            decoded_text = decoded_text[:-1] + '.'
+        else:
+            decoded_text += '.'
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:268]: ok

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -195,6 +308,12 @@ def set_n_grams(self, frequencies: dict) -> None:
         Args:
             frequencies (dict): Computed in advance frequencies for n-grams
         """
+        if not isinstance(frequencies, dict) or len(frequencies) == 0:
+            return None
+
+        self._n_gram_frequencies = frequencies
+
+        return None
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:316]: what exactly should I change?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -221,9 +360,25 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
...
+            if n_gram[:-1] == context:
+                token = n_gram[-1]
+                if token not in token_frequencies:
+                    token_frequencies[token] = freq
+
+        return token_frequencies
 
     def _extract_n_grams(
-        self, encoded_corpus: tuple[int, ...]
+            self, encoded_corpus: tuple[int, ...]
```
[the remark this answers is not in the document]
Comment (velgarita) [reply lab_3_generate_by_ngrams/main.py:381]: done

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -208,6 +325,26 @@ def build(self) -> int:
...
+        if not n_grams or not isinstance(n_grams, tuple):
+            return 1
+
+        context_freq_dict = {}
+
+        for n_gram in set(n_grams):
+            absolute_freq = n_grams.count(n_gram)
+            context = n_gram[:-1]
+            if context not in context_freq_dict:
+                context_freq_dict[context] = sum(1 for n_gram_1 in n_grams
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:341]: count

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -76,6 +108,14 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int):
+            return None
+
+        for token, ident in self._storage.items():
```
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:113]: What about this comment?

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -41,6 +45,26 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+
+        text_words = text.lower().split()
+        tokens = []
+        for word in text_words:
+            len_prev = len(tokens)
+            for alpha in word:
+                if alpha.isalpha():
+                    tokens.append(alpha)
+            if len(tokens) > len_prev:
```
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:57]: and this?

**Pull request merged**